File: src/services/cases.ts
import { invoke } from "@tauri-apps/api/core";

// ✅ Case Type Definition
export interface Case {
  file_id: number;
  case_number: string;
  purpose: "Ruling" | "Judgement" | "Other";
  uploaded_by: number;
  current_location: string;
  notes: string;
  date_recieved: string;
  required_on: string;
  required_on_signature: string;
  date_returned?: string | null;
  date_returned_signature?: string | null;
  deleted: number;
}

// ✅ Get All Cases
export async function getAllCases(): Promise<Case[]> {
  try {
    const cases: Case[] = await invoke("get_all_cases");
    return cases;
  } catch (error) {
    console.error("Error fetching cases:", error);
    throw error;
  }
}

// ✅ Create a Case
export async function createCase(
  data: Omit<
    Case,
    "case_id" | "date_returned" | "date_returned_signature" | "deleted"
  >
): Promise<{ message: string; status: string; case_id: number }> {
  try {
    const response = await invoke("create_case", {
      ...data,
    });
    console.log("Case created successfully:", response);
    return response as { message: string; status: string; case_id: number };
  } catch (error) {
    console.error("Error creating case:", error);
    throw error;
  }
}

// ✅ Get Case by ID
export async function getCase(caseId: number): Promise<Case | null> {
  try {
    const caseData: Case = await invoke("get_case", { case_id: caseId });
    return caseData;
  } catch (error) {
    console.error("Error fetching case:", error);
    return null;
  }
}

// ✅ Update Full Case Record
export async function updateCase(
  caseId: number,
  updates: Partial<Omit<Case, "case_id">>
): Promise<string> {
  try {
    const response: string = await invoke("update_case", {
      case_id: caseId,
      ...updates,
    });
    console.log("Case updated successfully:", response);
    return response;
  } catch (error) {
    console.error("Error updating case:", error);
    throw error;
  }
}

// ✅ Update Case Status (if applicable)
export async function updateCaseStatus(caseId: number, newStatus: string): Promise<string> {
  try {
    const response: string = await invoke("update_case_status", {
      case_id: caseId,
      new_status: newStatus,
    });
    console.log("Case status updated:", response);
    return response;
  } catch (error) {
    console.error("Error updating case status:", error);
    throw error;
  }
}

// ✅ Assign Staff to a Case (optional, if you still use assigned_staff_id)
export async function assignStaffToCase(caseId: number, staffId: number | null): Promise<string> {
  try {
    const response: string = await invoke("assign_staff_to_case", {
      case_id: caseId,
      staff_id: staffId,
    });
    console.log("Staff assigned to case:", response);
    return response;
  } catch (error) {
    console.error("Error assigning staff to case:", error);
    throw error;
  }
}

// ✅ Delete Case
export async function deleteCase(caseId: number): Promise<string> {
  try {
    const response: string = await invoke("delete_case", { case_id: caseId });
    console.log("Case deleted:", response);
    return response;
  } catch (error) {
    console.error("Error deleting case:", error);
    throw error;
  }
}

File: src-tauri/src/methods/cases.rs
use rusqlite::params;
use rusqlite::OptionalExtension;
use serde::{Deserialize, Serialize};
use serde_json::json;
use tauri::State;

use crate::AppState; // Import AppState

#[derive(Serialize, Deserialize)]
pub struct Case {
    pub case_id: i32,
    pub title: String,
    pub status: String,
    pub assigned_staff_id: Option<i32>,
    pub priority: String,
    pub date_created: String,
}

// ✅ Get All Cases
#[tauri::command]
pub fn get_all_cases(state: State<AppState>) -> Result<Vec<Case>, String> {
    let conn = state.conn.lock().unwrap();
    let mut stmt = conn
        .prepare(
            "SELECT case_id, title, status, assigned_staff_id, priority, date_created FROM cases",
        )
        .map_err(|e| format!("Failed to prepare statement: {}", e))?;

    let case_iter = stmt
        .query_map([], |row| {
            Ok(Case {
                case_id: row.get(0)?,
                title: row.get(1)?,
                status: row.get(2)?,
                assigned_staff_id: row.get(3)?,
                priority: row.get(4)?,
                date_created: row.get(5)?,
            })
        })
        .map_err(|e| format!("Failed to fetch cases: {}", e))?;

    let cases: Vec<Case> = case_iter.filter_map(Result::ok).collect();

    Ok(cases)
}

// ✅ Create a Case
#[tauri::command]
pub fn create_case(
    state: State<AppState>,
    title: String,
    assigned_staff_id: Option<i32>,
    priority: String,
) -> Result<serde_json::Value, String> {
    let conn = state.conn.lock().unwrap();

    match conn.execute(
        "INSERT INTO cases (title, status, assigned_staff_id, priority) VALUES (?1, 'Open', ?2, ?3)",
        params![title, assigned_staff_id, priority],
    ) {
        Ok(_) => {
            let new_case_id = conn.last_insert_rowid();
            Ok(json!({
                "message": "Case created successfully",
                "status": "success",
                "case_id": new_case_id
            }))
        }
        Err(e) => Err(format!("Failed to create case: {}", e)),
    }
}

// ✅ Get Case by ID
#[tauri::command]
pub fn get_case(state: State<AppState>, case_id: i32) -> Result<Option<Case>, String> {
    let conn = state.conn.lock().unwrap();
    let mut stmt = conn
        .prepare("SELECT case_id, title, status, assigned_staff_id, priority, date_created FROM cases WHERE case_id = ?1")
        .map_err(|e| format!("Failed to prepare statement: {}", e))?;

    let case = stmt
        .query_row(params![case_id], |row| {
            Ok(Case {
                case_id: row.get(0)?,
                title: row.get(1)?,
                status: row.get(2)?,
                assigned_staff_id: row.get(3)?,
                priority: row.get(4)?,
                date_created: row.get(5)?,
            })
        })
        .optional()
        .map_err(|e| format!("Failed to fetch case: {}", e))?;

    Ok(case)
}

// ✅ Update Case Status
#[tauri::command]
pub fn update_case_status(
    state: State<AppState>,
    case_id: i32,
    new_status: String,
) -> Result<String, String> {
    let conn = state.conn.lock().unwrap();
    match conn.execute(
        "UPDATE cases SET status = ?1 WHERE case_id = ?2",
        params![new_status, case_id],
    ) {
        Ok(_) => Ok("Case status updated successfully".to_string()),
        Err(e) => Err(format!("Failed to update case status: {}", e)),
    }
}

// ✅ Assign Staff to Case
#[tauri::command]
pub fn assign_staff_to_case(
    state: State<AppState>,
    case_id: i32,
    staff_id: Option<i32>,
) -> Result<String, String> {
    let conn = state.conn.lock().unwrap();
    match conn.execute(
        "UPDATE cases SET assigned_staff_id = ?1 WHERE case_id = ?2",
        params![staff_id, case_id],
    ) {
        Ok(_) => Ok("Staff assigned to case successfully".to_string()),
        Err(e) => Err(format!("Failed to assign staff to case: {}", e)),
    }
}

// ✅ Delete Case
#[tauri::command]
pub fn delete_case(state: State<AppState>, case_id: i32) -> Result<String, String> {
    let conn = state.conn.lock().unwrap();
    match conn.execute("DELETE FROM cases WHERE case_id = ?1", params![case_id]) {
        Ok(_) => Ok("Case deleted successfully".to_string()),
        Err(e) => Err(format!("Failed to delete case: {}", e)),
    }
}
